### ISSUE TITLE
Add a callback to support ocapus

### ISSUE THREAD

**Issue by edenlightning** [PyTorchLightning/pytorch-lightning#3305]
Opacus is a library that enables training PyTorch models with differential privacy. It supports training with minimal code changes required on the client, has little impact on training performance and allows the client to online track the privacy budget expended at any given moment.

https://github.com/pytorch/opacus

**Comment by lezwon** [comment]
This looks cool. I can work on this :]

**Comment by lezwon** [comment]
Have unassigned myself as I'm a little short on bandwidth rt. Can take it up sometime later if its still pending. :)

**Comment by cyr0930** [comment]
It could be a nice feature

**Comment by romovpa** [comment]
Hi everyone! I'm working on Lightning <> Opacus integration. [This PR](https://github.com/pytorch/opacus/pull/250) is a working prototype. 

Also I've implemented [a prototype with a Callback](https://github.com/romovpa/opacus/pull/5) that adds differential privacy to the training process. However, I'm at loss to come up with an elegant way of implementing this for our new API. 

In the new Opacus API we introduce classes DPOptimizer, DPModule (GradSampleModule) and DPDataLoader. To enable DP training one needs to wrap their model, optimizer and data_loader with those classes. Is there a way to replace the model and the optimizers with the wrapped ones during training in a Callback? 

Any help and comments are much appreciated.
@edenlightning @lezwon @cyr0930

**Comment by romovpa** [comment]
What is the best way to extend `Trainer.fit()` so that in case of differential privacy turned on, before running the loop it will wrap the model, optimizers and data_loader? 

In fact, to add DP, before training loop we call [PrivacyEngine.make_private()](https://github.com/pytorch/opacus/blob/d2f5948c177cd70c748befc713422e21c55979a4/opacus/privacy_engine.py#L94):
```
dp_model, dp_optimizer, dp_dataloader = privacy_engine.make_private(model, optimizer, dataloader)
```
After that, `dp_*` should be used instead of the original ones.
